singleuser auth as server extension

minrk opened the pull request:
This is a development branch, testing JupyterHub singleuser auth as a Jupyter Server extension

Jupyter Server 2.0 introduces two new APIs:

- IdentityProvider for configuring authentication
- Authorizer for configuring authorization (enables granular access, such as read-only)

This PR aims to implement these APIs for JupyterHub, without (almost) any use of monkeypatching or private APIs.
So far, I've got ~everything working in manual tests, aside from disable_user_config. I'm not 100% sure how feasible that's going to be, but hopefully I can manage.

It doesn't remove or change SingleUserNotebookApp. It mostly _copies_ code from there (essentially a fork of the singleuser code), so that the old code can be deprecated and removed at a later time.

Co-developing this with https://github.com/jupyter-server/jupyter_server/pull/825 as a way to validate that everything JupyterHub needs to fit in a server extension is covered.

TODO:

- [x] complete implementation
- [x] test/demonstrate _custom_ Authorizer that isn't the default JupyterHub Authorizer to make sure this is possible
- [x] add extension + server 2.0 to test matrix
- [x] figure out what to do for SingleUserNotebookApp, which is likely to be incompatible with server 2.0, and defaults in jupyterhub.singleuser
- [x] figure out disable_user_config
- [x] default to extension when running with Jupyter Server 2

minrk commented:
Finally got this working! (writing before CI is green, but I'm optimistic)

Key changes:

- the only change to the existing monkeypatch-app code is loading the custom page.html from a shared file
- In testing, the StubSingleUserServer which ran a singleuser server in a thread for easier assertions/introspection was too hard to keep working in all possible combinations. So I elected for a simpler, cleaner, more realistic launch of subprocess servers, and an extension to enable inspecting it over the network. This works fine and is a lot simpler.

The rest is just implementing the same behavior via standard server-2.0 public APIs instead of the monkeypatching.

Questions:

- should the extension be the _default_ if jupyter-server 2 is present? I think maybe yes. The only reason not to is Server 2 is out, so this is a change in behavior (I had hoped to finish this PR while 2 was in prerelease, but life got in the way).

consideRatio reviewed (commented):
Amazing work on this @minrk! I think it seems very thoroughly done!

My review notes mostly include details:
- request for additional details in docstrings
- questions about TODO notes
- small refactoring considerations for readability
- questions about things I don't understand

---

I've not reviewed things below line ~260 or so in the extension.py file, a bit into the code that you describe is copied from the mixin - everything else though!

minrk commented on jupyterhub/singleuser/__init__.py:
```diff
@@ -2,17 +2,44 @@
+    from .extension import main
+else:
+    _as_extension = False
+    try:
+        from .app import SingleUserNotebookApp, main
+    except ImportError:
+        # check for Jupyter Server 2.0 ?
+        from .extension import main
+    else:
+        # backward-compatibility
+        JupyterHubLoginHandler = SingleUserNotebookApp.login_handler_class
+        JupyterHubLogoutHandler = SingleUserNotebookApp.logout_handler_class
+        OAuthCallbackHandler = SingleUserNotebookApp.oauth_callback_handler_class
```
It's not really for readability. I generally try to only put the bit of code that I expect could fail within `try` blocks, that way there's no way I'll accidentally catch errors I don't mean to. That's not really possible in this particular case as assignments can't trigger ImportErrors (unless there are some real shenanigans going on), but is the pattern I try to follow generally.

In my head it goes like this:

1. try the import
2. if I don't 
3. if I have it, do this
4. if I don't, do

minrk commented on jupyterhub/singleuser/extension.py:
```diff
@@ -0,0 +1,659 @@
+        AsyncHTTPClient.configure(None, defaults={"ssl_options": ssl_context})
+        return AsyncHTTPClient()
+
+    async def check_hub_version(self):
```
Good catch! It should be used in the extension as well.

minrk commented:
@consideRatio thanks for the thoughtful review! I'll work on addressing the points now.

minrk commented on jupyterhub/tests/test_singleuser.py:
```diff
@@ -161,6 +166,27 @@ async def test_disable_user_config(app):
+        path = os.path.realpath(os.path.abspath(path))
+        assert not path.startswith(norm_home + os.path.sep)
+
+    # TODO: check legacy notebook config
+    # nbextensions_path
+    # static_custom_path
```
Yes, I think these things should be checked. I'll add that.

minrk commented:
had to re-sync with main to get the updated black config

minrk commented on jupyterhub/singleuser/templates/page.html:
```diff
@@ -0,0 +1,44 @@
+{% extends "templates/page.html" %} {% block header_buttons %} {{super()}}
```
added

minrk commented on jupyterhub/singleuser/extension.py:
```diff
@@ -0,0 +1,659 @@
+            yield p
+
+
+class JupyterHubLogoutHandler(LogoutHandler):
+    def get(self):
+        hub_auth = self.identity_provider.hub_auth
+        # clear single-user cookie
+        hub_auth.clear_cookie(self)
+        # redirect to hub to clear the rest
+        self.redirect(hub_auth.hub_host + url_path_join(hub_auth.hub_prefix, "logout"))
```
I've added a small clarification that `hub_auth` is responsible for the cookie. I don't like too many links to specific docs sections from code because they easily go stale and aren't ever checked. I think keywords and description that can be fed to search are more durable and reliable.

minrk commented on jupyterhub/singleuser/extension.py:
```diff
@@ -0,0 +1,659 @@
+class JupyterHubAuthorizer(Authorizer):
+    """Authorizer that looks for permissions in JupyterHub scopes"""
+
+    # TODO: https://github.com/jupyter-server/jupyter_server/pull/830
+    hub_auth = Instance(HubOAuth)
```
Updated. The Authorizer didn't need its own handle, since it should now refer to `self.identity_provider.hub_auth`

minrk commented:
@consideRatio thanks for the suggestions! I've added detail where I could in docstrings and comments based on your review. I also updated the extension to be usd by default with Jupyter Server 2, which for JupyterHub 4 I think is the right thing to do.

consideRatio commented:
The new changes looks great to me! I really love the docstring additions!!!

There is a timeout failure in the tests to resolve, but from what I could tell this looked good!

minrk commented:
Because this has been going for so long, I'm going to do the default-switch in its own PR, rather than tack on another change. I realize changing the default also means changing the test matrix to make sure the cases are still covered.

minrk commented:
@consideRatio good catch on the `check_hub_version`! Adding the call that should have been there is the reason for the hang, because the URL attribute needed to be updated for the extension, since it's on a different object. Should be okay now.

consideRatio reviewed (approved):
Wow amazing work on this!!! Super happy about how this is now an extension of jupyter-server!
